src/System.Text.Json/System/Text/Json/JsonReader.cs

```diff
@@ -1,383 +1,836 @@
+                        int startIndex = (int)Unsafe.ByteOffset(ref _buffer.DangerousGetPinnableReference(), ref src);
+
+                        // consumed is in characters, but our buffer is in bytes, so we need to double it for buffer slicing.
+                        int bytesConsumed = consumed << 1;
+
+                        Value = _buffer.Slice(startIndex, bytesConsumed);
+                        ValueType = JsonValueType.Number;
+                        return bytesConsumed;
+                    }
+                }
+            }
+            else
+                throw new NotImplementedException();
         }
 
-        public Utf8String GetValue()
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int ConsumeNaN(ref byte src, int count)
         {
-            var type = GetJsonValueType();
-            SkipEmpty();
-            switch (type)
+            Value = ReadOnlySpan<byte>.Empty;
+            ValueType = JsonValueType.NaN;
+
+            if (UseFastUtf8)
             {
-                case JsonValueType.String:
-                    return ReadStringValue();
-                case JsonValueType.Number:
-                    return ReadNumberValue();
-                case JsonValueType.True:
-                    return ReadTrueValue();
-                case JsonValueType.False:
-                    return ReadFalseValue();
-                case JsonValueType.Null:
-                    return ReadNullValue();
-                case JsonValueType.Object:
-                case JsonValueType.Array:
-                    return Utf8String.Empty;
-                default:
-                    throw new ArgumentException("Invalid json value type '" + type + "'.");
+                if (count < 3
+                    || Unsafe.Add(ref src, 0) != 'N'
+                    || Unsafe.Add(ref src, 1) != 'a'
+                    || Unsafe.Add(ref src, 2) != 'N')
+                {
+                    throw new JsonReaderException();
+                }
+
+                return 3;
+            }
+            else if (UseFastUtf16)
+            {
+                ref char chars = ref Unsafe.As<byte, char>(ref src);
+                if (count < 6
+                    || Unsafe.Add(ref chars, 0) != 'N'
+                    || Unsafe.Add(ref chars, 1) != 'a'
+                    || Unsafe.Add(ref chars, 2) != 'N')
+                {
+                    throw new JsonReaderException();
+                }
+
+                return 6;
             }
+            else
+                throw new NotImplementedException();
         }
 
-        private Utf8String ReadStringValue()
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int ConsumeNull(ref byte src, int count)
         {
-            _index++;
-            var count = _index;
-            do
+            Value = ReadOnlySpan<byte>.Empty;
+            ValueType = JsonValueType.Null;
+
+            if (UseFastUtf8)
```

Comment:
Yes, I thought about that. Ahson and I were talking about it yesterday. That has its own challenges though.